rust/libceed/src/vector.rs

```diff
@@ -375,6 +398,80 @@ impl<'a> Vector<'a> {
         self.check_error(ierr)
     }
 
+    /// Wrap a mutable slice in a Vector of the same length
+    ///
+    /// # arguments
+    ///
+    /// * `slice` - values to wrap in self; length must match
+    ///
+    /// ```
+    /// # use libceed::prelude::*;
+    /// # fn main() -> libceed::Result<()> {
+    /// # let ceed = libceed::Ceed::default_init();
+    /// let mut vec = ceed.vector(4)?;
+    /// let mut array = [10., 11., 12., 13.];
+    ///
+    /// {
+    ///     // `wrapper` holds a mutable reference to the wrapped slice
+    ///     //   that is dropped when `wrapper` goes out of scope
+    ///     let wrapper = vec.wrap_slice_mut(&mut array)?;
+    ///     vec.view()?.iter().enumerate().for_each(|(i, v)| {
+    ///         assert_eq!(*v, 10. + i as Scalar, "Slice not set correctly");
+    ///     });
+    ///
+    ///     // This line will not compile, as the `wrapper` holds mutable
+    ///     //   access to the `array`
+    ///     // array[0] = 5.0;
+    ///
+    ///     // Changes here are reflected in the `array`
+    ///     vec.set_value(5.0)?;
+    ///     vec.view()?.iter().for_each(|v| {
+    ///         assert_eq!(*v, 5.0 as Scalar, "Value not set correctly");
+    ///     });
+    /// }
+    ///
+    /// // 'array' remains changed
+    /// array.iter().for_each(|v| {
+    ///     assert_eq!(*v, 5.0 as Scalar, "Array not mutated correctly");
+    /// });
+    ///
+    /// // While changes to `vec` no longer affect `array`
+    /// vec.set_value(6.0)?;
+    /// array.iter().for_each(|v| {
+    ///     assert_eq!(*v, 5.0 as Scalar, "Array mutated without permission");
+    /// });
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn wrap_slice_mut<'b>(
+        &mut self,
+        slice: &'b mut [crate::Scalar],
+    ) -> crate::Result<VectorSliceWrapper<'b>> {
+        assert_eq!(self.length(), slice.len());
+        let (host, copy_mode) = (
+            crate::MemType::Host as bind_ceed::CeedMemType,
+            crate::CopyMode::UsePointer as bind_ceed::CeedCopyMode,
+        );
+        let ierr = unsafe {
+            bind_ceed::CeedVectorSetArray(
+                self.ptr,
+                host,
+                copy_mode,
+                slice.as_ptr() as *mut crate::Scalar,
+            )
+        };
+        self.check_error(ierr)?;
+
+        let mut ptr_copy = std::ptr::null_mut();
+        let ierr = unsafe { bind_ceed::CeedVectorReferenceCopy(self.ptr, &mut ptr_copy) };
+        self.check_error(ierr)?;
```

Comment:
The reference is because I'm making and then dropping a Vector, but I could avoid that by just keeping a raw C pointer instead of a C pointer converted to a Rust Vector

Reply:
No strong preference from me.